viewer/src/components/Commands.vue

```diff
@@ -333,11 +335,23 @@ export default class Commands extends Vue {
                 label = `Place a ${buildingName(building)}`;
               }
 
+              const createButtons = (command: string) => {
+                const coord = command.split(" ")[2];
+                const hexes = new Map([[this.engine.map.getS(coord), { coordinates: coord }]]);
+                this.$store.commit("gaiaViewer/highlightHexes", hexes);
+                return this.engine.round == Round.None ?
+                  [{
+                    command: coord,
+                    label: `Confirm Placement ${coord}`,
+                  } as ButtonData] : undefined;
+              };
```

Comment:
This code with highlight hexes is implementation details, it should be moved as much as possible in `MoveButton`.

There's no special need to repeat the placement in the confirm button, as far as I'm aware, it will still be displayed in the title like for boosters.

I would rather set `needConfirm` to true / add `buttons: [{command: '', label: 'Confirm mine'}]` and modifiy `MoveButton` to make it work.

Reply:
I couldn't get that to work. The coordinates are simply lost if we don't capture them while creating the button. Feel free to give it a try :smile:

Reply:
Here, but it could probably use some refactoring (the part in `subscribeHexClick`)

```patch
diff --git a/viewer/src/components/Commands.vue b/viewer/src/components/Commands.vue
index c9e0335..d97eed9 100644
--- a/viewer/src/components/Commands.vue
+++ b/viewer/src/components/Commands.vue
@@ -75,6 +75,7 @@ import Engine, {
   GaiaHex,
   Resource,
   Reward,
+  Round,
   SpaceMap,
   SubPhase,
   tiles,
@@ -317,6 +318,8 @@ export default class Commands extends Vue {
           const buildings = command.data.buildings as AvailableBuilding[];
           for (const building of Object.values(Building)) {
             const coordinates = buildings.filter((bld) => bld.building === building);
+            const buttons =
+              this.engine.round === Round.None ? [{ command: "", label: `Confirm ${buildingName(building)}` }] : null;
 
             if (coordinates.length > 0) {
               let label = `Build a ${buildingName(building)}`;
@@ -338,6 +341,8 @@ export default class Commands extends Vue {
                 command: `${Command.Build} ${building}`,
                 automatic: command.data.automatic,
                 hexes: hexMap(this.engine, coordinates),
+                buttons,
+                needConfirm: buttons?.length > 0,
               });
             }
           }
diff --git a/viewer/src/components/MoveButton.vue b/viewer/src/components/MoveButton.vue
index ad318c7..cba530d 100644
--- a/viewer/src/components/MoveButton.vue
+++ b/viewer/src/components/MoveButton.vue
@@ -94,7 +94,10 @@ export default class MoveButton extends Vue {
   }
 
   subscribeHexClick(callback: (hex: GaiaHex, highlight: HighlightHex) => void) {
-    this.subscribe("hexClick", payload => {
+    this.subscribe("hexClick", (payload) => {
+      if (this.button.needConfirm) {
+        this.button.buttons[0].command = payload.hex;
+      }
       callback(payload.hex, payload.highlight);
     });
   }
```

Reply:
Here's another patch - better refactored:

```patch
diff --git a/viewer/src/components/Commands.vue b/viewer/src/components/Commands.vue
index c9e0335..d97eed9 100644
--- a/viewer/src/components/Commands.vue
+++ b/viewer/src/components/Commands.vue
@@ -75,6 +75,7 @@ import Engine, {
   GaiaHex,
   Resource,
   Reward,
+  Round,
   SpaceMap,
   SubPhase,
   tiles,
@@ -317,6 +318,8 @@ export default class Commands extends Vue {
           const buildings = command.data.buildings as AvailableBuilding[];
           for (const building of Object.values(Building)) {
             const coordinates = buildings.filter((bld) => bld.building === building);
+            const buttons =
+              this.engine.round === Round.None ? [{ command: "", label: `Confirm ${buildingName(building)}` }] : null;
 
             if (coordinates.length > 0) {
               let label = `Build a ${buildingName(building)}`;
@@ -338,6 +341,8 @@ export default class Commands extends Vue {
                 command: `${Command.Build} ${building}`,
                 automatic: command.data.automatic,
                 hexes: hexMap(this.engine, coordinates),
+                buttons,
+                needConfirm: buttons?.length > 0,
               });
             }
           }
@@ -255,9 +254,16 @@ export default class MoveButton extends Vue {
     this.emitCommand();
   }
 
-  emitCommand(append?: string, params?: { disappear?: boolean; final?: boolean; times?: number; warnings?: BuildWarning[] }) {
+  emitCommand(
+    append?: string,
+    params?: { disappear?: boolean; final?: boolean; times?: number; warnings?: BuildWarning[] }
+  ) {
     console.log("emit command", this.button.command, append);
 
+    if (this.button.needConfirm && append) {
+      this.button.buttons[0].command = append;
+    }
+
     params = Object.assign({}, { disappear: true, final: false, times: 1 }, params);
```

Reply:
thanks a lot - I like that much better :smile: